lib/Sema/CSRanking.cpp

```diff
@@ -209,8 +209,16 @@ static bool isNominallySuperclassOf(Type type1, Type type2) {
 /// Determine the relationship between the self types of the given declaration
 /// contexts..
 static std::pair<SelfTypeRelationship, Optional<ProtocolConformanceRef>>
-computeSelfTypeRelationship(TypeChecker &tc, DeclContext *dc, DeclContext *dc1,
-                            DeclContext *dc2) {
+computeSelfTypeRelationship(TypeChecker &tc, DeclContext *dc, ValueDecl *decl1,
+                            ValueDecl *decl2) {
+  // If both declarations are operators, even through they
+  // might have Self such types are unrelated.
+  if (decl1->isOperator() && decl2->isOperator())
```

Comment:
It seems a bit odd to have this based on whether these are operator decls. Is that really the salient issue, or is it that operators declared in types are static?

Reply:
According to @DougGregor it's salient issue with operators, we shouldn't create additional "equals" constraints to match self types.

Reply:
Operators are the only thing in the language where we end up comparing members to non-members, and they have the additional (intentional!) restriction that at least one of the parameters must effectively be `Self`. So, I think this approach is reasonable.

Reply:
SGTM.